src/algebra/ring/projections.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.ring.basic
+
+/-!
+# Idempotents
+
+This file defines idempotents for an arbitary multiplication and proves some basic results:
+
+* In a semigroup, the product of two commuting idempotents is an idempotent;
```

Comment:
Another good result to have would be that `p ^ n` is an idempotent if `p` is.

Reply:
We don't have powers for non-unital multiplication yet do we?

Reply:
No, that would need to be in a `monoid` section

Reply:
@eric-wieser  I guess that I should start with something like:
```
variables {M : Type*} [has_mul M]

def nppow_rec : ℕ+ → M → M
| 1     a := a
| (n+1) a := a * nppow_rec n a
```
But this fails with:
```
non-exhaustive match, the following cases are missing:
nppow_rec M ⟨0, _⟩ _
```
Why is it trying to match case `0`?

Reply:
Let's not do too much at once here. Can you just add the lemma I suggest for `monoid`s? Defining a power operator for non-unital magmas is definitely out of scope for this PR.

Reply:
Okay, how would I exclude the `n=0` case from the induction and start at `n=1`?

Reply:
The equation compiler doesn't support it in a nice way, but you can use `pnat.rec_on` directly:
```
def nppow_rec (m : M) (n : ℕ+) : M :=
pnat.rec_on n (by exact m) (λ n, (*) m)
```
Please don't do that in this PR though.

Reply:
I've proved `p^(n + 1) = p` for p an idempotent in a monoid an n in ℕ. That may not be what you want though?